compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -224,15 +225,24 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
                 // Bools in union fields needs to be truncated.
                 *llval = bx.to_immediate(*llval, field);
                 // HACK(eddyb) have to bitcast pointers until LLVM removes pointee types.
-                *llval = bx.bitcast(*llval, bx.cx().immediate_backend_type(field));
+                let ty = bx.cx().immediate_backend_type(field);
+                if bx.type_kind(ty) == TypeKind::Pointer {
+                    *llval = bx.pointercast(*llval, ty);
+                }
```

Comment:
This removes the illegal `bitcast <4 x i32> to [4 x i32]`, which is good--but it doesn't replace that illegal conversion with a legal one, so we end up returning a value of type `<4 x i32>` when extracting a field of type `[4 x i32]`.

The repro example happens to work anyways because the next instruction is a store, and both `store <4 x i32>` and `store [4 x i32]` have the same effect. But it won't work if we use that value by value. (This may never happen, because I don't think we ever pass LLVM array types by value, but it's technically incorrect.)

To handle this fully, I think you'd have to special-case this kind of newtype field extraction (self `Abi::Vector`, field `Abi::Aggregate`) and roundtrip the vector through memory (alloca, `store <4 x i32>`, `load [4 x i32]`).

(I thought there might be [endianness concerns](https://llvm.org/docs/LangRef.html#vector-type), but it seems like that's [only a problem](https://godbolt.org/z/18o6nEcv5) for not-a-multiple-of-i8 types like i4)

Reply:
Good point! Split out the case for newtype of simd array